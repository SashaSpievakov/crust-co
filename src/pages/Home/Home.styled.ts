import styled from 'styled-components';

const Title = styled.h2`
  margin: 35px 0;

  @media screen and (max-width: 650px) {
    margin: 20px 0;
  }
`;

const Top = styled.section`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 30px;

  @media screen and (max-width: 900px) {
    flex-direction: column;
    align-items: start;
    margin-bottom: 15px;
  }
`;

const Block = styled(Top)`
  flex-direction: row;
  align-items: center;

  @media screen and (max-width: 900px) {
    margin-bottom: 50px;
  }

  @media screen and (max-width: 650px) {
    flex-direction: column;
    align-items: start;
  }
`;

const ErrorHeading = styled.h2`
  font-size: 32px;
  font-weight: 700;
  margin-bottom: 10px;

  @media screen and (max-width: 450px) {
    font-size: 25px;
  }
`;

const ErrorParagraph = styled.p`
  font-size: 22px;

  @media screen and (max-width: 650px) {
    font-size: 18px;
  }

  @media screen and (max-width: 450px) {
    font-size: 16px;
  }
`;

export { Title, Top, Block, ErrorHeading, ErrorParagraph };
