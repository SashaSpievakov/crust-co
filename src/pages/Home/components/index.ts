export { Categories } from './Categories';
export { ProductsContainer } from './ProductsContainer';
