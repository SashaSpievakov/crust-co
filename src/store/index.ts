export * from './slices';
export type { AppDispatch, AppStore, RootState } from './store';
export { setupStore } from './store';
