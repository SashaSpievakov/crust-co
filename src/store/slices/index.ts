export type { CartSliceState } from './cart';
export { addItem, clearCart, removeItem, removeItems } from './cart';
export { selectIsLight, setTheme } from './theme';
