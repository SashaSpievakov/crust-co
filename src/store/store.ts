import {
  combineReducers,
  configureStore,
  PreloadedState,
} from '@reduxjs/toolkit';

import categorySlice from './slices/category/reducer/categoryReducer';
import sortSlice from './slices/sort/reducer/sortReducer';
import searchSlice from './slices/search/reducer/searchValueReducer';
import theme from './slices/theme/reducer/themeReducer';
import cart from './slices/cart/reducer/cartReducer';
import itemAPI from '../services/ItemService';
import pizzasAPI from '../services/PizzasService';

const rootReducer = combineReducers({
  activeCategory: categorySlice,
  activeSort: sortSlice,
  searchValue: searchSlice,
  theme,
  cart,
  [itemAPI.reducerPath]: itemAPI.reducer,
  [pizzasAPI.reducerPath]: pizzasAPI.reducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware()
        .concat(itemAPI.middleware)
        .concat(pizzasAPI.middleware),
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
