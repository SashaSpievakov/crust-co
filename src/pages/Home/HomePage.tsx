import { ChangeEvent, FC, useEffect, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import { ErrorRequest } from 'src/components';
import { useDebounce } from 'src/hooks';
import { Container } from 'src/styles/Base.styled';

import { DropdownSelect, SearchItems } from '../../components/UI';
import { IPizzaItem } from '../../models';
import { pizzasAPI } from '../../services';
import { modifySearchParamsName } from '../../utils/modifySearchParamsName';
import { Categories, ProductsContainer } from './components';
import { Block, Title, Top } from './HomePage.styled';

export const sortNamesArr: string[] = ['rating', 'price', 'A to Z'];

export const HomePage: FC = () => {
  const [searchParams, setSearchParams] = useSearchParams();
  const [pizzas, setPizzas] = useState<IPizzaItem[]>([]);
  const [searchValue, setSearchValue] = useState<string>('');
  const debouncedSearchValue = useDebounce(searchValue, 100);
  const [activeSort, setActiveSort] = useState<number>(
    parseInt(searchParams.get('sortBy') || '0', 10) || 0,
  );
  const [activeCategory, setActiveCategory] = useState<number>(
    parseInt(searchParams.get('category') || '0', 10) || 0,
  );

  const sortSearchParam = modifySearchParamsName(sortNamesArr[activeSort]);
  const { data, isSuccess, isError, isLoading } = pizzasAPI.useFetchPizzasQuery(
    {
      activeCategory,
      sortSearchParam,
    },
  );

  useEffect(() => {
    document.title = 'Crust & Co.';
  }, []);

  useEffect(() => {
    if (isSuccess) {
      setPizzas(data);
    }
  }, [activeCategory, sortSearchParam, data, isSuccess]);

  const handleSortSelect = (index: number): void => {
    setActiveSort(index);
    setSearchParams({
      sortBy: index.toString(),
      category: activeCategory.toString(),
    });
  };

  const handleCategorySelect = (index: number): void => {
    setActiveCategory(index);
    setSearchParams({
      category: index.toString(),
      sortBy: activeSort.toString(),
    });
  };

  const onChangeSearch = (e: ChangeEvent<HTMLInputElement>): void => {
    setSearchValue(e.target.value);
  };

  const onClickClose = () => {
    setSearchValue('');
  };

  return isError ? (
    <ErrorRequest />
  ) : (
    <Container data-testid="homePage">
      <Top>
        <Categories
          activeCategory={activeCategory}
          onSelect={handleCategorySelect}
        />
        <DropdownSelect
          chosenValue={sortNamesArr[activeSort]}
          sortNamesArr={sortNamesArr}
          onSelect={handleSortSelect}
        />
      </Top>
      <Block>
        <Title>All Pizzas</Title>
        <SearchItems
          value={searchValue}
          width="240px"
          onChange={onChangeSearch}
          onClickClose={onClickClose}
          expandedWidth="300px"
        />
      </Block>
      <ProductsContainer
        isLoading={isLoading}
        items={pizzas}
        searchValue={debouncedSearchValue}
      />
    </Container>
  );
};
