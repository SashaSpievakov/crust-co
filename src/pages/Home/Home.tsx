import { useEffect, useState } from 'react';

import { Title, Block, Top, Error } from '../../styles/Base.styled';
import { selectSort } from '../../store/slices/sortSlice';
import { selectCategory } from '../../store/slices/categorySlice';
import Categories from '../../components/Categories/Categories';
import Sort from '../../components/Sort/Sort';
import SearchItems from '../../components/SearchItems/SearchItems';
import { IPizzaItem } from '../../models/IPizzaItem';
import { useAppSelector } from '../../hooks/reduxHooks';
import pizzasAPI from '../../services/PizzasService';
import ProductsContainer from '../../components/ProductsContainer/ProductsContainer';

export const sortNamesArr: string[] = ['rating', 'price', 'A to Z'];

const Home = () => {
  const [pizzas, setPizzas] = useState<IPizzaItem[]>([]);
  const activeCategory = useAppSelector(selectCategory);
  const activeSort = useAppSelector(selectSort);

  const sortedActiveName = sortNamesArr[activeSort];

  const sortPropertyName = (property: string) => {
    if (property === 'A to Z') {
      return 'name&order=asc';
    }
    return `${property}&order=desc`;
  };

  const sortedPropertyName = sortPropertyName(sortedActiveName);

  const { data, isSuccess, isError, isLoading } = pizzasAPI.useFetchPizzasQuery(
    {
      activeCategory,
      sortedPropertyName,
    },
  );

  useEffect(() => {
    if (isSuccess) {
      setPizzas(data);
    }
  }, [activeCategory, sortedPropertyName, data, isSuccess]);

  return (
    <>
      <Top>
        <Categories />
        <Sort sortNamesArr={sortNamesArr} />
      </Top>
      <Block>
        <Title>All Pizzas</Title>
        <SearchItems />
      </Block>
      {isError ? (
        <Error>
          <h2>Request Error</h2>
          <p>Coudn&apos;t get store items. Try your request again later.</p>
        </Error>
      ) : (
        <ProductsContainer isLoading={isLoading} items={pizzas} />
      )}
    </>
  );
};

export default Home;
