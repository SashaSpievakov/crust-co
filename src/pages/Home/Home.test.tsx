import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import { rest } from 'msw';

import pizzasAPI from '../../services/PizzasService';
import { setupStore } from '../../store/store';
import renderWithProviders from '../../tests/helpers/renderWithProviders';
import server from '../../tests/mocks/api/server';
import Home from './Home';

describe('Home Tests', () => {
  beforeAll(() => {
    server.listen();
  });

  afterAll(() => {
    server.close();
  });

  afterEach(() => {
    server.resetHandlers();
    setupStore().dispatch(pizzasAPI.util.resetApiState());
  });

  test('renders the Home page', async () => {
    renderWithProviders(<Home />);
    const lastItem = await screen.findByText(/diablo/i); // eslint-disable-line @typescript-eslint/no-unused-vars

    const loadedHome = screen.getByTestId('homePage');
    expect(loadedHome).toMatchSnapshot();
  });

  test('renders the Home page with a server requesr error', async () => {
    server.use(rest.get('*', (_req, res, ctx) => res.once(ctx.status(500))));
    renderWithProviders(<Home />);

    expect(
      await screen.findByText(/failed to get data from the server/i),
    ).toBeInTheDocument();
  });
});
