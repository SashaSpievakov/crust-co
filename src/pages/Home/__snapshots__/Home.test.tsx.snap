// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Tests renders the Home page 1`] = `
<article
  class="sc-bcXHqe bZXhLi"
  data-testid="homePage"
>
  <section
    class="sc-jSUZER eANtvp"
  >
    <section>
      <ul
        class="sc-iBYQkv gRQnbh"
      >
        <li
          aria-current="true"
          class="sc-ftTHYK hvoGON"
          tabindex="0"
        >
          All
        </li>
        <li
          aria-current="false"
          class="sc-ftTHYK hskZzN"
          tabindex="0"
        >
          Meat
        </li>
        <li
          aria-current="false"
          class="sc-ftTHYK hskZzN"
          tabindex="0"
        >
          Vegetarian
        </li>
        <li
          aria-current="false"
          class="sc-ftTHYK hskZzN"
          tabindex="0"
        >
          Grill
        </li>
        <li
          aria-current="false"
          class="sc-ftTHYK hskZzN"
          tabindex="0"
        >
          Spicy
        </li>
      </ul>
    </section>
    <section
      class="sc-pyfCe dcrHaX"
    >
      <div
        class="sc-jrcTuL bjUGsr"
        tabindex="0"
      >
        <svg
          class="sc-ipEyDJ kYhfAR"
          data-testid="popupIcon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.247 11.14 2.451 5.658C1.885 5.013 2.345 4 3.204 4h9.592a1 1 0 0 1 .753 1.659l-4.796 5.48a1 1 0 0 1-1.506 0z"
          />
        </svg>
        <b>
          Sort by
        </b>
        <span
          data-testid="currenChosenSortName"
        >
          rating
        </span>
      </div>
      <section
        aria-hidden="true"
        class="sc-csuSiG lmkHwq"
      >
        <ul
          aria-hidden="true"
          data-testid="popupList"
        >
          <li
            class="sc-eDWCr iBTyEZ"
            tabindex="0"
          >
            rating
          </li>
          <li
            class="sc-eDWCr cgIKVK"
            tabindex="0"
          >
            price
          </li>
          <li
            class="sc-eDWCr cgIKVK"
            tabindex="0"
          >
            A to Z
          </li>
        </ul>
      </section>
    </section>
  </section>
  <section
    class="sc-jSUZER sc-gKPRtg eANtvp eljnoa"
  >
    <h2
      class="sc-eDvSVe dtdrzn"
    >
      All Pizzas
    </h2>
    <section
      class="sc-bqWxrE dJvvDl"
    >
      <svg
        class="sc-ksBlkl jYsCes"
        data-testid="searchIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
        />
      </svg>
      <input
        class="sc-hBxehG iIJYd"
        placeholder="Search..."
        value=""
      />
    </section>
  </section>
  <section
    class="sc-eeMvmM fLEVJU"
  >
    <article
      class="sc-bYMpWt fjKzCa"
    >
      <a
        class="sc-kMjNwy jcOtpt"
        href="/item/3"
      >
        <picture>
          <source
            srcset="./assets/img/pizza3.webp"
            type="image/webp"
          />
          <source
            srcset="./assets/img/pizza3.png"
            type="image/png"
          />
          <img
            alt="Shrimp Pizza pizza"
            class="sc-ilhmMj hVMhhC"
            height="250"
            src="./assets/img/pizza3.png"
            width="250"
          />
        </picture>
        <h3
          class="sc-jIRcFI kqJrCP"
        >
          Shrimp Pizza
        </h3>
      </a>
      <section
        class="sc-gikAfH jGYPJX"
      >
        <ul>
          <li
            aria-current="true"
            class="sc-ezOQGI fAVUlZ"
            tabindex="0"
          >
            traditional
          </li>
          <li
            aria-current="false"
            class="sc-ezOQGI ezJDtU"
            tabindex="0"
          >
            thin
          </li>
        </ul>
        <ul>
          <li
            aria-current="true"
            class="sc-ezOQGI fAVUlZ"
            tabindex="0"
          >
            12
             inch
          </li>
          <li
            aria-current="false"
            class="sc-ezOQGI ezJDtU"
            tabindex="0"
          >
            14
             inch
          </li>
          <li
            aria-current="false"
            class="sc-ezOQGI ezJDtU"
            tabindex="0"
          >
            16
             inch
          </li>
        </ul>
      </section>
      <section
        class="sc-hhOBVt drvpMz"
      >
        <h3
          class="sc-gYbzsP eKShMY"
        >
          10
          $
        </h3>
        <button
          class="sc-fEXmlR sc-hHTYSt sc-kgTSHT cCXveJ goSwuU euaXrP"
        >
          <svg
            class="sc-iJnaPW hntdDm"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
            />
          </svg>
          <span>
            Add
          </span>
        </button>
      </section>
    </article>
    <article
      class="sc-bYMpWt fjKzCa"
    >
      <a
        class="sc-kMjNwy jcOtpt"
        href="/item/2"
      >
        <picture>
          <source
            srcset="./assets/img/pizza2.webp"
            type="image/webp"
          />
          <source
            srcset="./assets/img/pizza2.png"
            type="image/png"
          />
          <img
            alt="Pepperoni Pizza pizza"
            class="sc-ilhmMj hVMhhC"
            height="250"
            src="./assets/img/pizza2.png"
            width="250"
          />
        </picture>
        <h3
          class="sc-jIRcFI kqJrCP"
        >
          Pepperoni Pizza
        </h3>
      </a>
      <section
        class="sc-gikAfH jGYPJX"
      >
        <ul>
          <li
            aria-current="true"
            class="sc-ezOQGI fAVUlZ"
            tabindex="0"
          >
            traditional
          </li>
        </ul>
        <ul>
          <li
            aria-current="true"
            class="sc-ezOQGI fAVUlZ"
            tabindex="0"
          >
            12
             inch
          </li>
          <li
            aria-current="false"
            class="sc-ezOQGI ezJDtU"
            tabindex="0"
          >
            14
             inch
          </li>
        </ul>
      </section>
      <section
        class="sc-hhOBVt drvpMz"
      >
        <h3
          class="sc-gYbzsP eKShMY"
        >
          12
          $
        </h3>
        <button
          class="sc-fEXmlR sc-hHTYSt sc-kgTSHT cCXveJ goSwuU euaXrP"
        >
          <svg
            class="sc-iJnaPW hntdDm"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
            />
          </svg>
          <span>
            Add
          </span>
        </button>
      </section>
    </article>
    <article
      class="sc-bYMpWt fjKzCa"
    >
      <a
        class="sc-kMjNwy jcOtpt"
        href="/item/6"
      >
        <picture>
          <source
            srcset="./assets/img/pizza6.webp"
            type="image/webp"
          />
          <source
            srcset="./assets/img/pizza6.png"
            type="image/png"
          />
          <img
            alt="Diablo pizza"
            class="sc-ilhmMj hVMhhC"
            height="250"
            src="./assets/img/pizza6.png"
            width="250"
          />
        </picture>
        <h3
          class="sc-jIRcFI kqJrCP"
        >
          Diablo
        </h3>
      </a>
      <section
        class="sc-gikAfH jGYPJX"
      >
        <ul>
          <li
            aria-current="true"
            class="sc-ezOQGI fAVUlZ"
            tabindex="0"
          >
            traditional
          </li>
          <li
            aria-current="false"
            class="sc-ezOQGI ezJDtU"
            tabindex="0"
          >
            thin
          </li>
        </ul>
        <ul>
          <li
            aria-current="true"
            class="sc-ezOQGI fAVUlZ"
            tabindex="0"
          >
            12
             inch
          </li>
          <li
            aria-current="false"
            class="sc-ezOQGI ezJDtU"
            tabindex="0"
          >
            16
             inch
          </li>
        </ul>
      </section>
      <section
        class="sc-hhOBVt drvpMz"
      >
        <h3
          class="sc-gYbzsP eKShMY"
        >
          15
          $
        </h3>
        <button
          class="sc-fEXmlR sc-hHTYSt sc-kgTSHT cCXveJ goSwuU euaXrP"
        >
          <svg
            class="sc-iJnaPW hntdDm"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
            />
          </svg>
          <span>
            Add
          </span>
        </button>
      </section>
    </article>
  </section>
</article>
`;
